File: morning_exercises/day17.py
from __future__ import annotations

class BinaryNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

    @staticmethod
    def pathSum(root: BinaryNode | None, sum: int, targetSum: int)->bool:
        if root is None: return False

        if ((root.left is None) and (root.right is None)): return sum+root.val==targetSum

        return (BinaryNode.pathSum(root.left, sum+root.val, targetSum) or BinaryNode.pathSum(root.right, sum+root.val, targetSum))
    
root = BinaryNode(6)
root.left = BinaryNode(8)
root.right = BinaryNode(3)
root.left.left = BinaryNode(2)
root.left.right = BinaryNode(5)
root.right.left = BinaryNode(8)
root.right.right = BinaryNode(9)

print(BinaryNode.pathSum(root, 0, 18))

File: morning_exercises/day17.cpp
#include <ios>
#include <iostream>
#include <memory>
class BinaryNode {
    public:
        int val = 0;
        std::shared_ptr<BinaryNode> left = nullptr;
        std::shared_ptr<BinaryNode> right = nullptr;
    BinaryNode(int val): val(val) {}

    static bool pathSum(std::shared_ptr<BinaryNode> root, int sum, int targetSum) {
        if(root==nullptr) return false;

        if(root->left==nullptr && root->right==nullptr) return (sum+root->val==targetSum);
        return (pathSum(root->left, sum+root->val, targetSum) || pathSum(root->right, sum+root->val, targetSum));
    }
};

int main() {
    auto root = std::make_shared<BinaryNode>(5);
    root->left = std::make_shared<BinaryNode>(3);
    root->left->left = std::make_shared<BinaryNode>(1);
    root->left->right = std::make_shared<BinaryNode>(4);
    root->right = std::make_shared<BinaryNode>(2);
    root->right->right = std::make_shared<BinaryNode>(7);
    root->right->left = std::make_shared<BinaryNode>(1);
    
    std::cout << std::boolalpha;
    std::cout << BinaryNode::pathSum(root, 0, 69);
}